autometa/taxonomy/ncbi.py

```diff
@@ -364,28 +338,81 @@ def parse_merged(self):
             logger.debug('merged loaded')
         return merged
 
+    def get_lineage_names(self, taxid):
+        """
+        Returns the lineage names of the taxid
+
+        Parameters
+        ----------
+        taxid : int
+            identifer for a taxon in the Taxonomy Database by NCBI
+
+        Returns
+        -------
+        set
+            lineage of the taxid
+        """
+        lineage = set()
+        while taxid != 1:
+            if self.rank(taxid) not in NCBI.CANONICAL_RANKS:
+                taxid = self.parent(taxid)
+                continue
+            lineage.add(self.name(taxid))
+            taxid = self.parent(taxid)
+        return lineage
+
     def is_common_ancestor(self, parent_taxid, child_taxid):
-        """Determines whether the provided taxids have a non-root common ancestor
+        """
+        Determines whether the provided taxids have a non-root common ancestor
 
         Parameters
         ----------
         parent_taxid : int
-            Description of parameter `parent_taxid`.
+            identifer for a taxon in the Taxonomy Database by NCBI
         child_taxid : int
-            Description of parameter `child_taxid`.
+            identifer for a taxon in the Taxonomy Database by NCBI
 
         Returns
         -------
         boolean
-            True if taxids share a common ancestor and False otherwise
+            True if taxids share a common ancestor else False
         """
-        ancestor_taxid = child_taxid
-        while ancestor_taxid != 1:
-            if parent_taxid == ancestor_taxid:
-                return True
-            ancestor_taxid = self.parent(ancestor_taxid)
+        self.is_valid_taxid(parent_taxid)
+        self.is_valid_taxid(child_taxid)
+        parent_lineage = self.get_lineage_names(parent_taxid)
+        child_lineage = self.get_lineage_names(child_taxid)
+        common_ancestor = child_lineage.intersection(parent_lineage)
```

Comment:
OK, I have a few problems with this part of the code:

1. The `get_lineage_names` function gets the names of each of the taxids in the lineage of the two input taxids here. Names in the NCBI database are not unique, and there can be taxids with the same name and different rank, so it would be better to use the taxid itself rather than names, because they are guaranteed to be unique.
2. The `get_lineage_names` does not remove `'root'` from the canonical ranks, so this function is guaranteed to return `True` with any two valid taxids. 
3. Given the other two concerns, it is kind of pointless writing the new function `get_lineage_names` when we already have the `lineage` function - it is unnecessary bloat to the file.

Reply:
1. The script now uses taxid's for intersection and not `names`
2. I have removed `get_lineage_names`. Now the script uses the `lineage` function to get the taxids.
3. As per my understanding, the while loop prevents the `root` to be added to the list, as it terminates as soon as `taxid==1`. I verified this by printing the result as well.
```python
In [44]: from autometa.taxonomy.ncbi import NCBI                                 

In [45]: ncbi = NCBI("autometa/databases/ncbi/")                                 
DatabaseWarning: autometa/databases/ncbi/nr.gz needs to be formatted for diamond!

In [46]: ncbi.lineage(562)                                                       
Out[46]: 
[{'taxid': 562, 'name': 'escherichia coli', 'rank': 'species'},
 {'taxid': 561, 'name': 'escherichia', 'rank': 'genus'},
 {'taxid': 543, 'name': 'enterobacteriaceae', 'rank': 'family'},
 {'taxid': 91347, 'name': 'enterobacterales', 'rank': 'order'},
 {'taxid': 1236, 'name': 'gammaproteobacteria', 'rank': 'class'},
 {'taxid': 1224, 'name': 'proteobacteria', 'rank': 'phylum'},
 {'taxid': 2, 'name': 'bacteria', 'rank': 'superkingdom'}]
```

Reply:
Please correct me if I'm wrong, but the way I understood is that the `while taxid != 1` prevents the `root` to be added to the list (as the taxid of root is 1). My results on IDE:
```python
In [3]: from autometa.taxonomy.ncbi import NCBI                                 

In [4]: ncbi = NCBI("autometa/databases/ncbi/")                                 
DatabaseWarning: autometa/databases/ncbi/nr.gz needs to be formatted for diamond!

In [5]: ncbi.lineage(562)                                                       
Out[5]: 
[{'taxid': 562, 'name': 'escherichia coli', 'rank': 'species'},
 {'taxid': 561, 'name': 'escherichia', 'rank': 'genus'},
 {'taxid': 543, 'name': 'enterobacteriaceae', 'rank': 'family'},
 {'taxid': 91347, 'name': 'enterobacterales', 'rank': 'order'},
 {'taxid': 1236, 'name': 'gammaproteobacteria', 'rank': 'class'},
 {'taxid': 1224, 'name': 'proteobacteria', 'rank': 'phylum'},
 {'taxid': 2, 'name': 'bacteria', 'rank': 'superkingdom'},
 {'taxid': 131567, 'name': 'cellular organisms', 'rank': 'no rank'}]
```

However, when I ran it for different organisms I always got the 'cellular organisms' in the list, and thus the common ancestor was most of the time `True`. See here for what [cellular organisms](https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi) include.
I only got `False` when I tried with viruses. 
Is this what you were pointing at? Should I remove the entry of 'cellular organisms' ?

Reply:
Your problem there is that you should have set `canonical` to `True` when running `lineage`. For the rest of your comment, though. I can't see the old version of the `get_lineage_names` code any more. But you have to be careful with `while` loops. Yes, you are right that in the `lineage` function, this is the behavior you get (i.e. it doesn't add `root` to the lineage). However, remember that the while condition is checked BEFORE an iteration, so if during the iteration root is added to a datastructure, then what you are describing would not prevent root from being there. Does that make sense?

Reply:
I have now added a statement to discard '1' if present. However it can be my naive knowledge of programming that is preventing me from understanding how would taxid = 1 be added even with the `while` statement. 
Can be please give an example or point me towards a resource. Thanks :smile: